packages/runtime/lib/runtime.js

```diff
@@ -1884,45 +1921,155 @@ class Runtime extends EventEmitter {
+            throw new errors.InvalidArgumentError('maxHeapTotal', 'must be greater than 0')
+          }
+        } else {
+          throw new errors.InvalidArgumentError('maxHeapTotal', 'must be a number or a string representing a memory size')
+        }
+
+        if (currentHealth.maxHeapTotal === update.maxHeapTotal) {
+          this.logger.warn({ serviceId, maxHeapTotal: update.maxHeapTotal }, 'No change in the max heap total for service')
+          updateHeap = false
+        } else {
+          updateHeap = true
+        }
+      }
+
+      if (updateWorkers || updateHeap) {
+        validatedUpdates.push({ serviceId, config: serviceConfig, workers: update.workers, maxHeapTotal: update.maxHeapTotal, currentWorkers, currentHealth })
+      }
+    }
+
+    return validatedUpdates
+  }
+
+  async #updateServiceWorkersAndHeap (serviceId, config, serviceConfig, workers, maxHeapTotal, currentWorkers, currentHealth) {
+    if (currentWorkers > workers) {
+      // stop workers
+      await this.#updateServiceWorkers(serviceId, config, serviceConfig, workers, currentWorkers)
+      // update heap for current workers
+      await this.#updateServiceHeap(serviceId, config, serviceConfig, workers, currentHealth, maxHeapTotal)
+    } else {
+      // update service heap
+      await this.#updateServiceConfigHeap(serviceId, maxHeapTotal)
+      // start new workers with new heap
+      await this.#updateServiceWorkers(serviceId, config, serviceConfig, workers, currentWorkers)
+      // update heap for current workers
+      await this.#updateServiceHeap(serviceId, config, serviceConfig, currentWorkers, currentHealth, maxHeapTotal, false)
+    }
+  }
+
+  async #updateServiceHeap (serviceId, config, serviceConfig, currentWorkers, currentHealth, maxHeapTotal, updateConfig = true) {
+    let updated = 0
+    try {
+      if (updateConfig) {
+        await this.#updateServiceConfigHeap(serviceId, maxHeapTotal)
+      }
+
+      for (let i = 0; i < currentWorkers; i++) {
+        this.logger.info({ maxHeapTotal: { current: currentHealth.maxHeapTotal, new: maxHeapTotal } }, `Restarting service "${serviceId}" worker ${i} to update maxHeapTotal...`)
+
+        const worker = await this.#getWorkerById(serviceId, i)
+        worker[kConfig].health.maxHeapTotal = maxHeapTotal
+
+        await this.#replaceWorker(config, serviceConfig, currentWorkers, serviceId, i, worker)
+        updated++
+        this.logger.info({ maxHeapTotal: { current: currentHealth.maxHeapTotal, new: maxHeapTotal } }, `Restarted service "${serviceId}" worker ${i}`)
+      }
+    } catch (err) {
+      if (updated < 1) {
+        this.logger.error({ err }, 'Cannot update service maxHeapTotal, no worker updated')
+        await this.#updateServiceConfigHeap(serviceId, currentHealth.maxHeapTotal)
+      } else {
+        this.logger.error({ err }, `Cannot update service maxHeapTotal, updated workers: ${updated} out of ${currentWorkers}`)
+        // TODO in this case the value in
```

Comment:
@mcollina I don't have a good solution for this, any idea?

Reply:
In what situation can this happen? IMHO in this case we should just kill & restart all the services that could not be updated.